qutip/core/dimensions.py

```diff
@@ -352,6 +352,31 @@ def _frozen(*args, **kwargs):
     raise RuntimeError("Dimension cannot be modified.")
 
 
+def einsum(subscripts, *operands):
+    """
+    Implementation of numpy.einsum for Qobj.
+    Evaluates the Einstein summation convention on the operands.
+    Parameters
+    ----------
+    subscripts: str
+        Specifies the subscripts for summation as comma
+        separated list of subscript labels.
+    operands: list of array_like
+        These are the arrays for the operation.
+
+    Returns
+    -------
+    Qobj (numpy.complex128)
+        Result of einsum as Qobj (numpy.complex128 if result is scalar)
+    """
+    operands_array = [to_tensor_rep(op) for op in operands]
+    result = np.einsum(subscripts, *operands_array)
+    if result.shape == ():
+        return result
+    dims = [[d] for d in result.shape]
```

Comment:
Not sure if this work well with compound dimensions.
Could you add tests with multiple qubits systems.

Reply:
Sometime like 
`dims = [[d for d in result.shape[:result.ndim // 2], [d for d in result.shape[result.ndim // 2:]]`
would keep the compound dimensions.
We should raise warning or error if the output ndim is odd.